### PR TITLE
add each test result to EvaluateModelResponse

### PR DESCRIPTION
**Pull request by yuki-mt**
To visualize detailed evaluation result in drucker-dashboard, added more to EvaluateModelResponse.

The dashboard will show

- the test data is correct or not (= is_correct)
- and the prediction result (= prediction)
- with how much score (= predict_score)

for each test data.
(and detailed result (=option))

If it is approved, I will make PRs in drucker, drucker-dashboard and drucker-example repos along with the change as well.

### CODE REVIEW

**Comment by keigohtr** [comment]
Adding the details of prediction result is good. I basically agree with you. But,

```
    repeated bool is_correct = 6;
    repeated string prediction = 7;
    repeated float predict_score = 8;
```

- If these fields are required fields, the traffic volume is getting extremely huge compared with the current, and it means the big latency may occur. 
- The format of `repeated string prediction = 7;` cannot cover some cases. Sometimes a prediction result is not `string`.
- The format of `repeated float predict_score = 8;` cannot cover the case of multi-class classification.

### Suggestion
I think we need to design the use case at first.

- `EvaluateModel` uses for the visualization on Drucker dashboard, and also uses for automatic deployment after AB testing or canary testing. It means `EvaluateModel` must be light weight.
- The details of evaluation result are visualized on Drucker dashboard. If we want to evaluate instantly, we use `EvaluateModel`. If we want to analyze the details, we request to download the evaluation result data.
- The details of evaluation result must cover all return types of `DruckerWorker` (5 patterns).

I recommend you to develop the steps below. How do you think?

1. Create the new entry point which is for downloading the evaluation result.
1. When `EvaluateModel` is executed, Drucker saves the result data as a file to the directory which is the same place to the ML model. File name must be a combination of ML model version and evaluation data version.
1.  When the request of downloading evaluation data comes, Drucker sends the data and Drucker dashboard receives it.

**Comment by yuki-mt** [comment]
@keigohtr 
I agree with you.
So I added `EvaluateResult` to download evaluation data.

The request has `model_path` and `evaluate_ver`  (<- corresponding to `ML model version and evaluation data version.` in your comment)
If `evaluate_ver` is empty, DruckerDashboardServicer will return the newest result of the model evaluation.
If both are empty, DruckerDashboardServicer will return the newest result of any model evaluation.

As for response, each test data will be sent as bytes, and dashboard will manage how to decode it.
( `PredictInterface` will manage how to encode the data and how to decide the file name)

**Comment by yuki-mt** [comment]
@keigohtr 
Or can I try to receive test data asynchronously in `EvaluateModel` ?
The disadvantages of saving evaluation data as a file in drucker are

- drucker needs to manage more data,
- when displaying evaluation data,
    - huge data will be sent `drucker` -> `dashboard backend` -> `dashboard frontend` , which is time-consuming
    - If we use cache in `dashboard backend`, it can be more complicated.

So I think it is better if the both conditions are satisfied.

- EvaluateModel will return evaluation result summary (accuracy, precision...) very soon
- dashboard manages detailed evaluation result

(The document about gPRC Python Tutorial does not mention about receiving streaming data asynchronously, so I need to confirm it)

**Comment by keigohtr** [comment]
@yuki-mt 

> If evaluate_ver is empty, DruckerDashboardServicer will return the newest result of the model evaluation.
> If both are empty, DruckerDashboardServicer will return the newest result of any model evaluation.

In the future, since we will manage the evaluation dataset in Drucker dashboard, we can manage `evaluate_ver`. At the moment, it is enough to use a file name of evaluation data as `evaluate_ver`. So we don't need to care about these spec you mentioned.

**Comment by keigohtr** [comment]
@yuki-mt 
> As for response, each test data will be sent as bytes, and dashboard will manage how to decode it.

If we can manage how to decode it, it is the best.
However, since there are tons of machine learning module, I can't image the universal decoding system. So I think just showing data is enough at the moment.

Example
|correct/incorrect| data| top label| top score| labels...| scores...|

**Comment by keigohtr** [comment]
@yuki-mt 
> Or can I try to receive test data asynchronously in EvaluateModel ?

Good idea. I agree with you. But not `EvaluateModel` but `EvaluateResult`.
As I mentioned, I want to keep `EvaluateModel` as a light weight method. Saving the result to the storage is good idea since we never evaluate it twice.

Async transferring data is an excellent idea since we can check the data without waiting the completion of transference.

**Comment by yuki-mt** [comment]
@keigohtr 
I defined the following to define EvaluationDetail message.
(EvaluationDetail is detailed test result for each test data)

- SingleIO
- MultiIO
- IO
- SingleLabel
- MultiLabel
- Label

The reason why I divide IO and Label into single and multiple messages is that I want make [getting `is_single` like here](https://github.com/drucker/drucker/pull/4/files#diff-c6e29ead8807aed605a3a133fec25382R103) easy.

IO and Label can be used to create AnyInput and AnyOutput

File: drucker.proto [hunk 1]
```diff
@@ -145,15 +267,41 @@ message ModelResponse {
 // Request of EvaluateModel.
 message EvaluateModelRequest {
     bytes data = 1;
+    string data_path = 2;
```
Comment (yuki-mt) [inline drucker.proto:189]: This is where to save the test result in drucker.

File: drucker.proto [hunk 2]
```diff
@@ -115,6 +118,125 @@ service DruckerWorker {
...
+        ArrSfixed32 sfixed32_val = 9;
+        ArrSfixed64 sfixed64_val = 10;
+        ArrFloat float_val = 11;
+        ArrDouble double_val = 12;
+        ArrBool bool_val = 13;
+        ArrString string_val = 14;
+    }
+}
+
+message IO {
```
Comment (keigohtr) [inline drucker.proto:152]: This `message` is never used.

File: drucker.proto [hunk 3]
```diff
@@ -145,15 +267,41 @@ message ModelResponse {
...
+message EvaluationResultResponse {
+    EvaluationMetrics metrics = 1;
+
+    message Detail {
+        bytes data = 1;
+        bool is_correct = 2;
+        Label label = 3;
+    }
+
+    repeated Detail detail = 2;
```
Comment (keigohtr) [inline drucker.proto:225]: Does this mean that we can request some results in one request?

**Comment by yuki-mt** [comment]
@keigohtr I made `IO` simple. Please review it.

**Comment by keigohtr** [comment]
@yuki-mt 
At the memont, this spec is enough for evaluation. We will expand this to output `bytes` data.
LGTM.

**Review by keigohtr — APPROVED** [review]
LGTM

**Pull request merged**